File: scala/05-api/src/main/scala/x/Clickstream.scala
package x

/**
this will process JSON clickstream data

Usage:
spark-submit --class 'x.Clickstream'  target/scala-2.11/testapp_2.11-1.0.jar    <files to process>

*/


import org.apache.spark.SparkContext._
import org.apache.spark.sql.SparkSession

object Clickstream {
  def main(args: Array[String]) {
    if (args.length < 1) {
      println ("need file(s) to load")
      System.exit(1)
    }

    val input = args(0) // can be wildcard  'dir/*.log'

    val spark = SparkSession.builder().
                appName("Clicks -- MYNAME").
                getOrCreate()


    val clickstream = spark.read.json(input)

    val count =  clickstream.count
    println("### total clickstream records " + count)

    val domainCount = clickstream.groupBy("domain").count()

    domainCount.orderBy("count").show()
  }
}

File: scala/05-api/python/clickstream.py
# Usage:
# spark-submit  clickstream.py  <files to process>
# spark-submit  --master spark://localhost:7077 clickstream.py  <files to process>

## dealing with logs
# spark-submit  clickstream.py  <files to process>  2> logs
#

# Multiple files can be specified
#e.g:
#- with 4G executor memory and turning off verbose logging
#   spark-submit  --master spark://localhost:7077 --executor-memory 4g   --driver-class-path logging/    clickstream.py  /data/cickstream/json/
#

import sys
import time
from pyspark.sql import SparkSession

if len(sys.argv) < 2:
    sys.exit("need clickstream json to load")

spark = SparkSession.builder.appName("Process Clickstream -- MYNAME").getOrCreate()

f = sys.argv[1]
clickstream = spark.read.json(f)
count = clickstream.count()
print ("record count ", count)

domainCount = clickstream.groupBy("domain").count()
domainCount.orderBy("count", ascending=False).show()

#line = input('### Hit Ctrl+C to terminate the program...')
spark.stop()  # close the session
